distro/otel_test.go

```diff
@@ -611,54 +644,64 @@ func (coll *collector) Start(t *testing.T) {
 	require.NoError(t, conn.Close())
 }
 
-func (coll *collector) SpansExportRequest() *spansExportRequest {
-	// Give some time for the gRPC server to process the request.
-	timeout := time.NewTimer(100 * time.Millisecond)
-	defer timeout.Stop()
-	select {
-	case got := <-coll.traceService.requests:
-		return &got
-	case <-timeout.C:
-		return nil
-	}
+func (coll *collector) ExportedSpans() *spansExportRequest {
+	// stop to make sure all requests are processed and synchronized
+	coll.grpcSrv.GracefulStop()
+
+	defer coll.traceService.mtx.Unlock()
+	coll.traceService.mtx.Lock()
+	return coll.traceService.data
 }
 
-func (coll *collector) MetricsExportRequest() *metricsExportRequest {
-	// Give some time for the gRPC server to process the request.
-	timeout := time.NewTimer(100 * time.Millisecond)
-	defer timeout.Stop()
-	select {
-	case got := <-coll.metricsService.requests:
-		return &got
-	case <-timeout.C:
-		return nil
-	}
+func (coll *collector) ExportedMetrics() *metricsExportRequest {
+	// stop to make sure all requests are processed and synchronized
+	coll.grpcSrv.GracefulStop()
+
+	defer coll.metricsService.mtx.Unlock()
+	coll.metricsService.mtx.Lock()
+	return coll.metricsService.data
 }
 
 func (ctss *collectorTraceServiceServer) Export(ctx context.Context, exp *ctpb.ExportTraceServiceRequest) (*ctpb.ExportTraceServiceResponse, error) {
 	rs := exp.ResourceSpans[0]
-	scopeSpans := rs.ScopeSpans[0]
+
 	headers, _ := metadata.FromIncomingContext(ctx)
 
-	ctss.requests <- spansExportRequest{
-		Header:   headers,
-		Resource: rs.GetResource(),
-		Spans:    scopeSpans.GetSpans(),
+	ctss.mtx.Lock()
+	if ctss.data == nil {
+		// headers and resource should be the same. set them once
+		ctss.data = &spansExportRequest{
+			Header:   headers,
+			Resource: rs.GetResource(),
+		}
 	}
+	// concat all spans
+	for _, scopeSpans := range rs.ScopeSpans {
+		ctss.data.Spans = append(ctss.data.Spans, scopeSpans.GetSpans()...)
+	}
+
+	ctss.mtx.Unlock()
 
 	return &ctpb.ExportTraceServiceResponse{}, nil
 }
 
 func (cmss *collectorMetricsServiceServer) Export(ctx context.Context, exp *cmpb.ExportMetricsServiceRequest) (*cmpb.ExportMetricsServiceResponse, error) {
 	rs := exp.ResourceMetrics[0]
-	scopeMetrics := rs.ScopeMetrics[0]
 	headers, _ := metadata.FromIncomingContext(ctx)
 
-	cmss.requests <- metricsExportRequest{
-		Header:   headers,
-		Resource: rs.GetResource(),
-		Metrics:  scopeMetrics.GetMetrics(),
+	cmss.mtx.Lock()
+	if cmss.data == nil {
+		// headers and resource should be the same. set them once
+		cmss.data = &metricsExportRequest{
+			Header:   headers,
+			Resource: rs.GetResource(),
+		}
+	}
+	// concat all metrics
+	for _, scopeMetrics := range rs.ScopeMetrics {
+		cmss.data.Metrics = append(cmss.data.Metrics, scopeMetrics.GetMetrics()...)
 	}
+	cmss.mtx.Unlock()
```

Comment:
```suggestion
	cmss.mtx.Lock()
	defer cmss.mtx.Unlock()
	if cmss.data == nil {
		// headers and resource should be the same. set them once
		cmss.data = &metricsExportRequest{
			Header:   headers,
			Resource: rs.GetResource(),
		}
	}
	// concat all metrics
	for _, scopeMetrics := range rs.ScopeMetrics {
		cmss.data.Metrics = append(cmss.data.Metrics, scopeMetrics.GetMetrics()...)
	}
```

Reply:
part of https://github.com/signalfx/splunk-otel-go/pull/1848/commits/854a374398ef4a11b52fcaf03a1f9d6bfb9a63d1